File: golang/tests/pkg/integration/pty_integration_test.go
package integration_test

import (
	"strings"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/pty"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/agentbay/testutil"
)

const ptyImageID = "linux_latest"

func createPtySession(t *testing.T) (*agentbay.AgentBay, *agentbay.Session) {
	t.Helper()
	apiKey := testutil.GetTestAPIKey(t)
	ab, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay: %v", err)
	}
	params := agentbay.NewCreateSessionParams().WithImageId(ptyImageID)
	result, err := ab.Create(params)
	if err != nil {
		t.Fatalf("Error creating session: %v", err)
	}
	if !result.Success {
		t.Fatalf("Session creation failed")
	}
	t.Logf("Session created: %s", result.Session.SessionID)
	return ab, result.Session
}

func TestPtyCreateAndEcho(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	var chunks [][]byte
	handle, err := session.Pty.Create(pty.CreateOptions{
		OnData: func(data []byte) {
			chunks = append(chunks, append([]byte{}, data...))
		},
	})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}
	if !handle.IsConnected() {
		t.Fatal("handle should be connected")
	}

	time.Sleep(1 * time.Second)

	if err := handle.SendInput([]byte("echo 'PTY_GO_OK_54321'\r")); err != nil {
		t.Fatalf("SendInput failed: %v", err)
	}
	time.Sleep(2 * time.Second)

	combined := ""
	for _, c := range chunks {
		combined += string(c)
	}
	if !strings.Contains(combined, "PTY_GO_OK_54321") {
		t.Fatalf("Expected output to contain 'PTY_GO_OK_54321', got: %s", combined)
	}

	handle.Disconnect()
}

func TestPtyResize(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	var chunks [][]byte
	handle, err := session.Pty.Create(pty.CreateOptions{
		OnData: func(data []byte) {
			chunks = append(chunks, append([]byte{}, data...))
		},
	})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}

	time.Sleep(1 * time.Second)

	if err := handle.Resize(120, 40); err != nil {
		t.Fatalf("Resize failed: %v", err)
	}
	time.Sleep(1 * time.Second)

	if err := handle.SendInput([]byte("echo \"cols=$(tput cols) lines=$(tput lines)\"\r")); err != nil {
		t.Fatalf("SendInput failed: %v", err)
	}
	time.Sleep(2 * time.Second)

	combined := ""
	for _, c := range chunks {
		combined += string(c)
	}
	if !strings.Contains(combined, "cols=120") {
		t.Fatalf("Expected 'cols=120' in output, got: %s", combined)
	}

	handle.Disconnect()
}

func TestPtyList(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	handle, err := session.Pty.Create(pty.CreateOptions{})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}
	time.Sleep(1 * time.Second)

	sessions, err := session.Pty.List()
	if err != nil {
		t.Fatalf("Failed to list PTY sessions: %v", err)
	}

	found := false
	for _, s := range sessions {
		if s.PtySessionID == handle.PtySessionID() {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("Created PTY %s not found in list", handle.PtySessionID())
	}

	handle.Disconnect()
}

func TestPtyExitCode(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	handle, err := session.Pty.Create(pty.CreateOptions{})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}
	time.Sleep(1 * time.Second)

	if err := handle.SendInput([]byte("exit\r")); err != nil {
		t.Fatalf("SendInput failed: %v", err)
	}
	exitCode, err := handle.Wait(10000)
	if err != nil {
		t.Fatalf("Wait failed: %v", err)
	}

	t.Logf("Exit code: %d", exitCode)
	if handle.IsConnected() {
		t.Fatal("handle should be disconnected after exit")
	}
}

func TestPtyKill(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	handle, err := session.Pty.Create(pty.CreateOptions{})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}
	time.Sleep(1 * time.Second)

	if err := handle.Kill(); err != nil {
		t.Fatalf("Kill failed: %v", err)
	}
	exitCode, err := handle.Wait(10000)
	if err != nil {
		t.Fatalf("Wait failed: %v", err)
	}

	if exitCode != -9 {
		t.Fatalf("Expected exit code -9, got %d", exitCode)
	}
}

func TestPtyDisconnectReconnect(t *testing.T) {
	ab, session := createPtySession(t)
	defer func() {
		ab.Delete(session)
	}()

	handle1, err := session.Pty.Create(pty.CreateOptions{})
	if err != nil {
		t.Fatalf("Failed to create PTY: %v", err)
	}
	time.Sleep(1 * time.Second)

	ptyID := handle1.PtySessionID()
	handle1.Disconnect()
	if handle1.IsConnected() {
		t.Fatal("handle should be disconnected")
	}

	time.Sleep(1 * time.Second)

	var output2 [][]byte
	handle2, err := session.Pty.Connect(ptyID, func(data []byte) {
		output2 = append(output2, append([]byte{}, data...))
	})
	if err != nil {
		t.Fatalf("Connect failed: %v", err)
	}
	if !handle2.IsConnected() {
		t.Fatal("handle2 should be connected")
	}

	if err := handle2.SendInput([]byte("echo 'RECONNECTED_GO'\r")); err != nil {
		t.Fatalf("SendInput failed: %v", err)
	}
	time.Sleep(2 * time.Second)

	combined := ""
	for _, c := range output2 {
		combined += string(c)
	}
	if !strings.Contains(combined, "RECONNECTED_GO") {
		t.Fatalf("Expected 'RECONNECTED_GO' in output, got: %s", combined)
	}

	handle2.Disconnect()
}

File: typescript/tests/integration/pty.integration.test.ts
import { AgentBay, Session } from "../../src";
import { getTestApiKey } from "../utils/test-helpers";

const IMAGE_ID = "linux_latest";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

jest.setTimeout(120000);

describe("PTY Integration", () => {
  let agentBay: AgentBay;
  let session: Session;

  beforeAll(async () => {
    const apiKey = getTestApiKey();
    agentBay = new AgentBay({ apiKey });
    const result = await agentBay.create({ imageId: IMAGE_ID });
    expect(result.success).toBe(true);
    session = result.session!;
  });

  afterAll(async () => {
    try {
      if (session) await agentBay.delete(session);
    } catch (e) {
      console.warn("Failed to delete session:", e);
    }
  });

  test("create and echo", async () => {
    const chunks: Uint8Array[] = [];
    const handle = await session.pty.create({
      onData: (data) => chunks.push(data),
    });
    expect(handle.ptySessionId).toBeTruthy();
    expect(handle.isConnected).toBe(true);

    await sleep(1000);
    await handle.sendInput(new TextEncoder().encode("echo 'PTY_TS_OK_12345'\r"));
    await sleep(2000);

    const combined = Buffer.concat(chunks).toString("utf-8");
    expect(combined).toContain("PTY_TS_OK_12345");

    handle.disconnect();
  });

  test("ctrl+c interrupt", async () => {
    const chunks: Uint8Array[] = [];
    const handle = await session.pty.create({
      onData: (data) => chunks.push(data),
    });
    await sleep(1000);

    await handle.sendInput(new TextEncoder().encode("sleep 100\r"));
    await sleep(1000);

    await handle.sendInput(new Uint8Array([0x03]));
    await sleep(2000);

    const combined = Buffer.concat(chunks).toString("utf-8");
    expect(combined).toMatch(/\^C|\$/);

    handle.disconnect();
  });

  test("resize", async () => {
    const chunks: Uint8Array[] = [];
    const handle = await session.pty.create({
      onData: (data) => chunks.push(data),
    });
    await sleep(1000);

    await handle.resize(120, 40);
    await sleep(1000);

    await handle.sendInput(
      new TextEncoder().encode('echo "cols=$(tput cols) lines=$(tput lines)"\r')
    );
    await sleep(2000);

    const combined = Buffer.concat(chunks).toString("utf-8");
    expect(combined).toContain("cols=120");
    expect(combined).toContain("lines=40");

    handle.disconnect();
  });

  test("list sessions", async () => {
    const handle = await session.pty.create();
    await sleep(1000);

    const sessions = await session.pty.list();
    const ids = sessions.map((s) => s.ptySessionId);
    expect(ids).toContain(handle.ptySessionId);

    handle.disconnect();
  });

  test("disconnect and reconnect", async () => {
    const output1: Uint8Array[] = [];
    const handle1 = await session.pty.create({
      onData: (data) => output1.push(data),
    });
    await sleep(1000);

    await handle1.sendInput(
      new TextEncoder().encode("echo 'BEFORE_DISCONNECT'\r")
    );
    await sleep(1000);

    const ptyId = handle1.ptySessionId;
    handle1.disconnect();
    expect(handle1.isConnected).toBe(false);

    const output2: Uint8Array[] = [];
    const handle2 = await session.pty.connect(ptyId, (data) =>
      output2.push(data)
    );
    expect(handle2.isConnected).toBe(true);

    await handle2.sendInput(
      new TextEncoder().encode("echo 'AFTER_RECONNECT'\r")
    );
    await sleep(2000);

    const combined = Buffer.concat(output2).toString("utf-8");
    expect(combined).toContain("AFTER_RECONNECT");

    handle2.disconnect();
  });

  test("exit with exit code", async () => {
    const handle = await session.pty.create();
    await sleep(1000);

    await handle.sendInput(new TextEncoder().encode("exit\r"));
    const exitCode = await handle.wait(10000);
    expect(typeof exitCode).toBe("number");
    expect(handle.isConnected).toBe(false);
  });

  test("kill with exit code -9", async () => {
    const handle = await session.pty.create();
    await sleep(1000);

    await handle.kill();
    const exitCode = await handle.wait(10000);
    expect(exitCode).toBe(-9);
  });

  test("multiple ptys", async () => {
    const output1: Uint8Array[] = [];
    const output2: Uint8Array[] = [];
    const handle1 = await session.pty.create({
      onData: (data) => output1.push(data),
    });
    const handle2 = await session.pty.create({
      onData: (data) => output2.push(data),
    });
    await sleep(1000);

    await handle1.sendInput(new TextEncoder().encode("echo 'FROM_PTY_1'\r"));
    await handle2.sendInput(new TextEncoder().encode("echo 'FROM_PTY_2'\r"));
    await sleep(2000);

    const combined1 = Buffer.concat(output1).toString("utf-8");
    const combined2 = Buffer.concat(output2).toString("utf-8");
    expect(combined1).toContain("FROM_PTY_1");
    expect(combined2).toContain("FROM_PTY_2");

    handle1.disconnect();
    handle2.disconnect();
  });
});
